Run-time Error: Slice bounds out of range when querying history

## Expected Behavior

- Join server
- Observe broken threads
- Query for threads
- Receive broken threads until none exist

## Current Behavior

- Join server
- Observe broken threads
- Query for threads
- Crash with:

```
panic: runtime error: slice bounds out of range

goroutine 20 [running]:
github.com/arborchat/muscadine/archive.(*Archive).Needed(0xc00007a7b0, 0xa, 0x0, 0x0, 0x0)
	/Users/jwhetton/go/src/github.com/arborchat/muscadine/archive/archive.go:56 +0x2c3
github.com/arborchat/muscadine/tui.(*TUI).queryNeeded(0xc00012c050, 0xc00011a000, 0xc0001540f0, 0x28, 0x5)
	/Users/jwhetton/go/src/github.com/arborchat/muscadine/tui/tui.go:238 +0x45
github.com/arborchat/muscadine/tui.(*TUI).queryNeeded-fm(0xc00011a000, 0xc0001540f0, 0x7, 0xc00005e901)
	/Users/jwhetton/go/src/github.com/arborchat/muscadine/tui/keybindings.go:34 +0x3e
github.com/whereswaldon/gocui.(*Gui).execKeybindings(0xc00011a000, 0xc0001540f0, 0xc0001e7d70, 0x115d618, 0x0, 0x0)
	/Users/jwhetton/go/src/github.com/whereswaldon/gocui/gui.go:643 +0xbe
github.com/whereswaldon/gocui.(*Gui).onKey(0xc00011a000, 0xc0001e7d70, 0xc0000469d0, 0xc00007a9b0)
	/Users/jwhetton/go/src/github.com/whereswaldon/gocui/gui.go:607 +0x1b0
github.com/whereswaldon/gocui.(*Gui).handleEvent(0xc00011a000, 0xc0001e7d70, 0x2, 0x0)
	/Users/jwhetton/go/src/github.com/whereswaldon/gocui/gui.go:427 +0x40
github.com/whereswaldon/gocui.(*Gui).MainLoop(0xc00011a000, 0x11cfb4f, 0x4)
	/Users/jwhetton/go/src/github.com/whereswaldon/gocui/gui.go:373 +0x2c3
github.com/arborchat/muscadine/tui.(*TUI).mainLoop.func1(0xc0001260c0, 0xc00012c050)
	/Users/jwhetton/go/src/github.com/arborchat/muscadine/tui/tui.go:118 +0x320
created by github.com/arborchat/muscadine/tui.(*TUI).mainLoop
	/Users/jwhetton/go/src/github.com/arborchat/muscadine/tui/tui.go:104 +0x62
```

Crash occurs on first attempt to query for broken threads.

Can confirm
```
cameron@School-PC ~ [2]> muscadine -username /dev/urandom arbor.whetton.net:7777
panic: runtime error: slice bounds out of range

goroutine 6 [running]:
github.com/arborchat/muscadine/archive.(*Archive).Needed(0xc00008a930, 0xa, 0x0, 0x0, 0x0)
        /home/cameron/go/src/github.com/arborchat/muscadine/archive/archive.go:56 +0x2c3
github.com/arborchat/muscadine/tui.(*TUI).queryNeeded(0xc000118050, 0xc00013c000, 0xc00016e0f0, 0x28, 0x5)
        /home/cameron/go/src/github.com/arborchat/muscadine/tui/tui.go:237 +0x45
github.com/arborchat/muscadine/tui.(*TUI).queryNeeded-fm(0xc00013c000, 0xc00016e0f0, 0x7, 0xc000001b01)
        /home/cameron/go/src/github.com/arborchat/muscadine/tui/keybindings.go:33 +0x3e
github.com/jroimartin/gocui.(*Gui).execKeybindings(0xc00013c000, 0xc00016e0f0, 0xc000217d70, 0x5af838, 0x0, 0x0)
        /home/cameron/go/src/github.com/jroimartin/gocui/gui.go:629 +0xbe
github.com/jroimartin/gocui.(*Gui).onKey(0xc00013c000, 0xc000217d70, 0xc0000501a0, 0xc000152050)
        /home/cameron/go/src/github.com/jroimartin/gocui/gui.go:593 +0x1b0
github.com/jroimartin/gocui.(*Gui).handleEvent(0xc00013c000, 0xc000217d70, 0x2, 0x0)
        /home/cameron/go/src/github.com/jroimartin/gocui/gui.go:413 +0x40
github.com/jroimartin/gocui.(*Gui).MainLoop(0xc00013c000, 0x674c30, 0x4)
        /home/cameron/go/src/github.com/jroimartin/gocui/gui.go:373 +0x2c3
github.com/arborchat/muscadine/tui.(*TUI).mainLoop.func1(0xc00001c180, 0xc000118050)
        /home/cameron/go/src/github.com/arborchat/muscadine/tui/tui.go:117 +0x320
created by github.com/arborchat/muscadine/tui.(*TUI).mainLoop
        /home/cameron/go/src/github.com/arborchat/muscadine/tui/tui.go:103 +0x62
cameron@School-PC ~ [2]> 
```

I got the same error on the notifications branch. I think this information is important since the notifications branch is 13 commits behind master. The most notable difference is using jroimartin's gocui instead of Waldon's fork.
```
cameron@School-PC ~/g/s/g/a/muscadine (notifications) [2]> ./muscadine -username "caton101" redacted:7777
-999
panic: runtime error: slice bounds out of range

goroutine 9 [running]:
github.com/arborchat/muscadine/archive.(*Archive).Needed(0xc0000849c0, 0xa, 0x0, 0x0, 0x0)
        /home/cameron/go/src/github.com/arborchat/muscadine/archive/archive.go:56 +0x2c3
github.com/arborchat/muscadine/tui.(*TUI).queryNeeded(0xc0000b21e0, 0xc000102090, 0xc00013c000, 0x28, 0x5)
        /home/cameron/go/src/github.com/arborchat/muscadine/tui/tui.go:232 +0x45
github.com/arborchat/muscadine/tui.(*TUI).queryNeeded-fm(0xc000102090, 0xc00013c000, 0x7, 0xc000136001)
        /home/cameron/go/src/github.com/arborchat/muscadine/tui/tui.go:351 +0x3e
github.com/jroimartin/gocui.(*Gui).execKeybindings(0xc000102090, 0xc00013c000, 0xc0001d7e38, 0x5b1a48, 0x0, 0x0)
        /home/cameron/go/src/github.com/jroimartin/gocui/gui.go:629 +0xbe
github.com/jroimartin/gocui.(*Gui).onKey(0xc000102090, 0xc0001d7e38, 0xc000050960, 0xc000050800)
        /home/cameron/go/src/github.com/jroimartin/gocui/gui.go:593 +0x1b0
github.com/jroimartin/gocui.(*Gui).handleEvent(0xc000102090, 0xc0001d7e38, 0x2, 0x0)
        /home/cameron/go/src/github.com/jroimartin/gocui/gui.go:413 +0x40
github.com/jroimartin/gocui.(*Gui).MainLoop(0xc000102090, 0x0, 0x0)
        /home/cameron/go/src/github.com/jroimartin/gocui/gui.go:373 +0x2c3
github.com/arborchat/muscadine/tui.(*TUI).mainLoop.func1(0xc00001c5a0, 0xc0000b21e0)
        /home/cameron/go/src/github.com/arborchat/muscadine/tui/tui.go:112 +0x1d0
created by github.com/arborchat/muscadine/tui.(*TUI).mainLoop
        /home/cameron/go/src/github.com/arborchat/muscadine/tui/tui.go:102 +0x62
cameron@School-PC ~/g/s/g/a/muscadine (notifications) [2]> /home/cameron/go/src/github.com/jroimartin/gocui/gui.go:373
```
Ignore the `-999` line in the terminal output. It is part of my feature testing.

--- a/archive/archive.go
+++ b/archive/archive.go
@@ -53,7 +53,7 @@ func (a *Archive) Needed(n int) []string {
 	if n >= len(needed) {
 		return needed
 	}
-	return needed[len(a.chronological)-n:]
+	return needed[len(needed)-n:]
 }
 
 // Has returns whether the archive contains a message with the given ID.

--- a/archive/archive_test.go
+++ b/archive/archive_test.go
@@ -327,3 +327,35 @@ func TestNeeded(t *testing.T) {
 		}
 	}
 }
+
+// TestLongHistNeeded is a regression test that ensures that a very long message history with many unknown
+// parents doesn't crash the client. (github.com/arborchat/muscadine/issues/61)
+func TestLongHistNeeded(t *testing.T) {
+	g := gomega.NewGomegaWithT(t)
+	a := newOrSkip(t)
+	message := arbor.ChatMessage{
+		UUID:      "whatever",
+		Parent:    "something",
+		Content:   "a lame test",
+		Timestamp: 500000,
+		Username:  "Socrates",
+	}
+	// add lots of messages with known parents
+	for i := 0; i < 100; i++ {
+		message.Parent = message.UUID   // make a child of the previous iteration's message
+		message.UUID += "a"             // ensure new id each iteration
+		added := new(arbor.ChatMessage) // allocate new memory for message so all pointers don't go to same address
+		*added = message
+		addOrSkip(t, a, added)
+	}
+	// add ten messages with unknown parents
+	for i := 0; i < 10; i++ {
+		message.Parent += "b"           // make a child of the previous iteration's message
+		message.UUID += "a"             // ensure new id each iteration
+		added := new(arbor.ChatMessage) // allocate new memory for message so all pointers don't go to same address
+		*added = message
+		addOrSkip(t, a, added)
+	}
+	needed := a.Needed(5)
+	g.Expect(len(needed)).To(gomega.Equal(5))
+}

